### PR TITLE
Support GC and compression layer updates in remote storage sync

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Closes https://github.com/neondatabase/neon/issues/999
Does not address synchronisation between GC, compression and sync processes to avoid races on the local fs, that feels worth a separate issue and PR.
Deals with GC and compression accounting, by actually uploading and removing files that are created and deleted locally during GC and compression loops.
For that, the following steps were made:

1. https://github.com/neondatabase/neon/pull/1603/commits/13041834e7c7d42249b2a2a1fb1bb8ba7c63f39f 

* Layered repository holds onto `PathBuf` for historic layers to put it into all new places where sync tasks are submitted.
More such places added for GC and compaction.
* Timeline batching is reworked: no more [confusing](https://github.com/neondatabase/neon/pull/1518#discussion_r852154471) enum variants, more straightforward bacthing code.

2. https://github.com/neondatabase/neon/pull/1603/commits/cc42ed08556db3826279e86ae2eda9a90f4bd88f Layers that are created by compaction are now uploaded to S3, layers that are removed by GC and compaction are now deleted from the remote storage

3. https://github.com/neondatabase/neon/pull/1603/commits/82e40ffc0ce10f3f01713b21a9623f60d196d8a5 Sync queue got refactored into a `struct` instead of a set of static methods to encapsulate its batching logic and to test it properly.

### CODE REVIEW

**Comment by SomeoneToIgnore** [comment]
Ok, looks like the Arc removal commit causes two tests to fail, here are the runs for a separate branch with this commit: https://app.circleci.com/pipelines/github/neondatabase/neon/5920/workflows/1df3aa37-b5f1-44ec-98fa-fe78237d14f5

Both issues do not have visible panics in pageserver logs, so the investigation might take some time.
WDYT, is `Arc` & `dyn` removal worth following that path or should we slap yet another bandaid to `Layer` interface?

**Comment by SomeoneToIgnore** [comment]
I've removed the `Arc` fiddling after some further investigations, since did not find what's wrong with that code despite not changing much of it. That seem to fix the benchmarks, so should be good to review?

**Review by LizardWizzard — COMMENTED** [review]
Thanks for the patch! I've left some comments.

One other thought I have. I'm not sure in benefits of current batching schema. It is already quite complicated with limit of timelines per batch and underlying concurrency limit. Logic around composite tasks sometimes is pretty complex too. Not sure that other solutions can somehow help with complexity. Batching adds latency which will be a problem for on-demand download. Lets discuss that in person and try to extend docs with more details

It is possible that during deletion only some of the files will be actually deleted. I'm not sure that compaction code or gc can handle these cases. Let's have an issue for that to add such tests. Maybe @hlinakka can say if it will work right away or there will be some tweaks required.

For future let's keep tenantid -> tenant_id renames separate, sometimes they occupy significant portion of the diff. Maybe we need to run sed over the entire source code and forget about it :)

File: pageserver/src/layered_repository/layer_map.rs [hunk 1]
```diff
@@ -58,7 +59,14 @@ pub struct LayerMap {
...
     /// other 2D data structure would be the long-term solution here.
-    historic_layers: Vec<Arc<dyn Layer>>,
+    historic_layers: Vec<HistoricLayer>,
+}
+
+/// A representation of a layer that's on disk and tracked in pageserver.
+#[derive(Clone)]
+pub struct HistoricLayer {
+    pub local_path: PathBuf,
+    pub layer: Arc<dyn Layer>,
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository/layer_map.rs:69]: Nope, this is a file name, not actually a path.
Moreover, even the `InMemoryLayer` implements it and even returns something.

File: pageserver/src/remote_storage/storage_sync.rs [hunk 2]
```diff
@@ -658,126 +649,127 @@ where
...
-        ControlFlow::Continue(task) => task,
-        ControlFlow::Break(aborted_task) => {
-            match aborted_task {
-                SyncTask::Download(_) => {
-                    index
-                        .write()
+    let upload_data = batch.upload.clone();
+    let download_data = batch.download.clone();
+    // Run both upload and download tasks concurrently (not in parallel):
+    // download and upload tasks do not conflict and spoil the pageserver state even if they are executed in parallel.
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/remote_storage/storage_sync.rs:655]: https://docs.rs/tokio/latest/tokio/macro.join.html#runtime-characteristics

There are two ways to launch async tasks:
* `.await` and the whole async machinery underneath will be executed as a single task
* `tokio::spawn` and this task will be run separately, maybe even on a separate thread (hence more strict `Send + 'static` requirements on what exactly can be spawned). Moreover, panics in such cases will be wrapped and returned as a top-level result from the `.await` on the spawned task

In this case, what will happen is that we start running the futures in the order they're specified, so first starts the first upload, eventually it yields and runtime picks the next upload, etc., eventually it comes to downloads (maybe due to all uploads being completed, maybe due to all of them being suspended due to reasons).
If anything panics, the whole thing gets aborted, also blocking calls will halt the entire machinery.

While with `span` calls, if run separately for both tasks, it is something close to what `FuturesOrdered` do: after the first few polls, both tasks start their own awaiting machinery, independently concurrently working on uploads and downloads.

File: pageserver/src/remote_storage/storage_sync.rs [hunk 3]
```diff
@@ -1015,51 +1087,25 @@ where
...
-        SyncTask::DownloadAndUpload(download_task, _) if skip_upload => {
-            ControlFlow::Continue(SyncTask::Download(download_task))
-        }
-        SyncTask::DownloadAndUpload(_, upload_task) if skip_download => {
-            ControlFlow::Continue(SyncTask::Upload(upload_task))
-        }
-        not_skipped => ControlFlow::Continue(not_skipped),
+    if current_attempt > 0 {
+        let seconds_to_wait = 2.0_f64.powf(current_attempt as f64 - 1.0).min(30.0);
+        info!("Waiting {seconds_to_wait} seconds before starting the task");
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/remote_storage/storage_sync.rs:1105]: Yes, agree that it's not optimal.
On the other hand, not sure if the on demand tasks should go into the same queue: they are top proirity tasks and might even get their own separate runtime adjusted for a different workload.

File: pageserver/src/remote_storage/storage_sync.rs [hunk 4]
```diff
@@ -134,151 +143,123 @@ lazy_static! {
...
-            warn!("Failed to enqueue a sync task: the sender is not initialized");
-            false
+impl SyncQueue {
+    fn new(max_timelines_per_batch: NonZeroUsize) -> Self {
+        let (sender, receiver) = mpsc::unbounded_channel();
+        Self {
+            len: AtomicUsize::new(0),
+            max_timelines_per_batch,
+            sender,
+            receiver: Mutex::new(receiver),
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/remote_storage/storage_sync.rs:164]: This is an unfortunate combination of sync and async worlds.
The root cause is the fact that all receive methods on the `receiver` are `&mut self`, such as `try_recv(&mut self) -> ...` and the fact that we have to keep a set of static pub methods for submitting sync messages (upload/download/delete), ideally without blocking too much. 
Those static methods have to be synchronous, since are called from sync parts of pageserver.

For those static methods, a queue has to be placed into a shared state, we use `lazy_static` or `once_cell` and `Mutex/RwLock/RefCell` for interior mutability: `once_cell` allows to retrieve the data as `fn get(&self) -> Option<&T>`, lazy static also returns `&` and we need `&mut` on the channel.

But either those mutexes have to come from `std::sync` module or I have to add some static runtime to `.await` on the tokio ones.

----

Or I could push the async mutex down where it is now and call it in the `async` code during the task processing.

Any other, better ways? Maybe I should use a static runtime after all?

File: pageserver/src/remote_storage/storage_sync.rs [hunk 5]
```diff
@@ -134,151 +143,123 @@ lazy_static! {
...
-    /// Fetches a task batch, not bigger than the given limit.
-    /// Not blocking, can return fewer tasks if the queue does not contain enough.
-    /// Batch tasks are split by timelines, with all related tasks merged into one (download/upload)
-    /// or two (download and upload, if both were found in the queue during batch construction).
-    pub async fn next_task_batch(
-        receiver: &mut UnboundedReceiver<(ZTenantTimelineId, SyncTask)>,
-        max_timelines_to_sync: NonZeroUsize,
-    ) -> ControlFlow<(), HashMap<ZTenantTimelineId, SyncTask>> {
+    /// Fetches a task batch, getting every existing entry from it, grouping by timelines and merging the tasks for every timeline.
+    /// A timeline has to care to not to delete cetain layers from the remote storage before the corresponding uploads happen.
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/remote_storage/storage_sync.rs:180]: I can remove that doc line, don't see any ways to enforce with the current model, only to react on various events that happened already.

What we need to decide is how to deal with the local FS deletions during the tasks.
I think it's fine to reread a local metadata and submit attempt to update the index with it, at least now: we have to update the remote json file anyway, why not do that with the updated metadata then, keeping it more consistent.

We need more tests on that case yet.

File: pageserver/src/remote_storage/storage_sync.rs [hunk 6]
```diff
@@ -658,126 +649,127 @@ where
...
+                            .ok();
+                    }
+                }
+            }
+            None
+        }
+        .instrument(info_span!("download_timeline_data")),
+    );
+
+    if let Some(delete_data) = batch.delete {
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/remote_storage/storage_sync.rs:735]: Theoretically, why not?
Will add a removal, just in case.

File: pageserver/src/layered_repository.rs [hunk 7]
```diff
@@ -1834,11 +1824,23 @@ impl LayeredTimeline {
...
+            let mut layer_paths_to_upload = HashSet::with_capacity(partitioning.parts.len());
             for part in partitioning.parts.iter() {
                 if self.time_for_new_image_layer(part, lsn)? {
-                    self.create_image_layer(part, lsn)?;
+                    let new_path = self.create_image_layer(part, lsn)?;
+                    layer_paths_to_upload.insert(new_path);
                 }
             }
+            if self.upload_layers.load(atomic::Ordering::Relaxed) {
+                let metadata = load_metadata(self.conf, self.timeline_id, self.tenant_id)
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository.rs]: Yep, no need, thank you a lot for this mention, it helped me to solve the failing tests issue faster.

File: pageserver/src/remote_storage/storage_sync.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/remote_storage/storage_sync.rs:164]: > we have to keep a set of static pub methods for submitting sync messages

There is no need to access the receiver during submission, right? Are there any problems with decoupling the receiver from this struct entirely? Looks like we can return `(Self, Receiver)` from `SyncQueue::new` and then move the receiver to the sync thread without any synchronization requierd in the future

File: pageserver/src/remote_storage/storage_sync.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/remote_storage/storage_sync.rs:180]: I see the one single problematic case:
If there will be an upload that fails because compaction deleted the file. Will this job update metadata? If the answer is yes what metadata will it be? We need to maintain the invariant of remote consistent lsn which should grow without introducing any gaps. If the metadata we read is newer than the files we actually uploaded - we're in trouble. I think the safest option is to avoid the metadata update entirely. If there is a proper upload task with metadata snapshot taken at the time of a checkpoint then it is ok to update it

File: pageserver/src/remote_storage/storage_sync.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/remote_storage/storage_sync.rs:694]: Yeah, the async machinery underneath is clear. But it is still unclear what spoil means here. What is the possible problem here? Could you please elaborate on that

File: pageserver/src/remote_storage/storage_sync.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/remote_storage/storage_sync.rs:768]: So it is possible that we delete + download the same data. Which can make a surprise for gc/compaction. Most probably we don't have such tests. Maybe worth a todo or an issue. Feel free to resolve

**Review by LizardWizzard — COMMENTED** [review]
Looks good to me. There are still some unresolved questions around edge-cases, but nothing serious

File: pageserver/src/remote_storage/storage_sync.rs [hunk 8]
```diff
@@ -961,27 +994,37 @@ async fn upload_timeline_data<P, S>(
...
+            // This is "fine" due the asynchronous nature of the sync loop: it only reacts to events and might need to
+            // retry the upload tasks, if S3 or network is down: but during this time, pageserver might still operate and
+            // run compaction/gc threads, removing redundant files from disk.
+            // It's not good to pause GC/compaction because of those and we would rather skip such uploads.
+            //
+            // Yet absence of such files might also mean that the timeline metadata file was updated (GC moves the Lsn forward, for instance).
+            // So we read the local version of the metadata file and submit is during the upload: we do update the index with the upload data
+            // and we do update the remote json index part file with the uploaded paths anyway: adding to the same both places an updated
+            // metadata is fine too.
+            // If other uploads get scheduled between those events with older metadata, such
```
Comment (LizardWizzard) [inline pageserver/src/remote_storage/storage_sync.rs:1008]: Looks like something is missing :)

File: pageserver/src/remote_storage/storage_sync.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/remote_storage/storage_sync.rs:164]: That actually worked not that bad, even though a bit verbose, thanks.

File: pageserver/src/remote_storage/storage_sync.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/remote_storage/storage_sync.rs:180]: Yes, that's actually a very good point, I've thought completely oppositely first.
I've moved this comment to the place where we detect such missing file, removed the entire variant branch with this "soft" error.
